In [1]:
//Databrick spark xml is a replacement for edu.umd.cloud9.collections package that is used in the book. Cloud9 ceased being maintained in 2016
import $ivy.`com.databricks:spark-xml_2.12:0.7.0`
import com.databricks.spark.xml._
import $ivy.`org.apache.spark::spark-sql:2.4.3`
import org.apache.spark.sql._
import org.apache.log4j._

Downloaded https://repo1.maven.org/maven2/com/databricks/spark-xml_2.12/0.7.0/spark-xml_2.12-0.7.0.pom
Downloaded https://repo1.maven.org/maven2/commons-io/commons-io/2.6/commons-io-2.6.pom
Downloaded https://repo1.maven.org/maven2/org/apache/commons/commons-parent/42/commons-parent-42.pom
Downloaded https://repo1.maven.org/maven2/org/apache/apache/18/apache-18.pom
Downloaded https://repo1.maven.org/maven2/commons-io/commons-io/2.6/commons-io-2.6.jar
Downloaded https://repo1.maven.org/maven2/com/databricks/spark-xml_2.12/0.7.0/spark-xml_2.12-0.7.0-sources.jar
Downloaded https://repo1.maven.org/maven2/com/databricks/spark-xml_2.12/0.7.0/spark-xml_2.12-0.7.0.jar
Downloaded https://repo1.maven.org/maven2/commons-io/commons-io/2.6/commons-io-2.6-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-sql_2.12/2.4.3/spark-sql_2.12-2.4.3.pom
Downloaded https://repo1.maven.org/maven2/sh/almond/almond-spark_2.12/0.8.2/almond-spark_2.12-0.8.2.pom
Downloaded https://repo1.ma

Downloaded https://repo1.maven.org/maven2/org/scala-lang/scala-reflect/2.11.12/scala-reflect-2.11.12.pom
Downloaded https://repo1.maven.org/maven2/io/dropwizard/metrics/metrics-graphite/3.1.5/metrics-graphite-3.1.5.pom
Downloaded https://repo1.maven.org/maven2/com/twitter/chill-java/0.9.3/chill-java-0.9.3.pom
Downloaded https://repo1.maven.org/maven2/org/apache/commons/commons-crypto/1.0.0/commons-crypto-1.0.0.pom
Downloaded https://repo1.maven.org/maven2/com/vlkan/flatbuffers/1.2.0-3f79e055/flatbuffers-1.2.0-3f79e055.pom
Downloaded https://repo1.maven.org/maven2/com/ning/compress-lzf/1.0.3/compress-lzf-1.0.3.pom
Downloaded https://repo1.maven.org/maven2/joda-time/joda-time/2.9.9/joda-time-2.9.9.pom
Downloaded https://repo1.maven.org/maven2/commons-net/commons-net/3.1/commons-net-3.1.pom
Downloaded https://repo1.maven.org/maven2/org/apache/ivy/ivy/2.4.0/ivy-2.4.0.pom
Downloaded https://repo1.maven.org/maven2/javax/activation/activation/1.1.1/activation-1.1.1.pom
Downloaded https://repo

Downloaded https://repo1.maven.org/maven2/org/glassfish/jersey/core/jersey-common/2.22.2/jersey-common-2.22.2.pom
Downloaded https://repo1.maven.org/maven2/io/airlift/aircompressor/0.10/aircompressor-0.10.pom
Downloaded https://repo1.maven.org/maven2/com/google/protobuf/protobuf-java/2.5.0/protobuf-java-2.5.0.pom
Downloaded https://repo1.maven.org/maven2/org/roaringbitmap/RoaringBitmap/0.7.45/RoaringBitmap-0.7.45.pom
Downloaded https://repo1.maven.org/maven2/net/razorvine/pyrolite/4.13/pyrolite-4.13.pom
Downloaded https://repo1.maven.org/maven2/org/apache/parquet/parquet-encoding/1.10.1/parquet-encoding-1.10.1.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-launcher_2.12/2.4.3/spark-launcher_2.12-2.4.3.pom
Downloaded https://repo1.maven.org/maven2/org/apache/curator/curator-recipes/2.6.0/curator-recipes-2.6.0.pom
Downloaded https://repo1.maven.org/maven2/commons-codec/commons-codec/1.10/commons-codec-1.10.pom
Downloaded https://repo1.maven.org/maven2/org/apache/com

Downloaded https://repo1.maven.org/maven2/io/dropwizard/metrics/metrics-parent/3.1.5/metrics-parent-3.1.5.pom
Downloaded https://repo1.maven.org/maven2/org/eclipse/jetty/jetty-project/9.4.20.v20190813/jetty-project-9.4.20.v20190813.pom
Downloaded https://repo1.maven.org/maven2/io/netty/netty-parent/4.1.17.Final/netty-parent-4.1.17.Final.pom
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-project-dist/2.6.5/hadoop-project-dist-2.6.5.pom
Downloaded https://repo1.maven.org/maven2/org/roaringbitmap/RoaringBitmapParent/0.7.45/RoaringBitmapParent-0.7.45.pom
Downloaded https://repo1.maven.org/maven2/org/sonatype/oss/oss-parent/7/oss-parent-7.pom
Downloaded https://repo1.maven.org/maven2/org/apache/avro/avro-parent/1.8.2/avro-parent-1.8.2.pom
Downloaded https://repo1.maven.org/maven2/org/apache/apache/7/apache-7.pom
Downloaded https://repo1.maven.org/maven2/org/glassfish/jersey/containers/project/2.22.2/project-2.22.2.pom
Downloaded https://repo1.maven.org/maven2/com/google/

Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-common/2.6.5/hadoop-common-2.6.5.pom
Downloaded https://repo1.maven.org/maven2/org/json4s/json4s-core_2.12/3.5.3/json4s-core_2.12-3.5.3.pom
Downloaded https://repo1.maven.org/maven2/org/scala-lang/scala-reflect/2.12.8/scala-reflect-2.12.8.pom
Downloaded https://repo1.maven.org/maven2/org/eclipse/jetty/jetty-io/9.4.20.v20190813/jetty-io-9.4.20.v20190813.pom
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/external/javax.inject/2.4.0-b34/javax.inject-2.4.0-b34.pom
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/osgi-resource-locator/1.0.1/osgi-resource-locator-1.0.1.pom
Downloaded https://repo1.maven.org/maven2/org/glassfish/jersey/media/jersey-media-jaxb/2.22.2/jersey-media-jaxb-2.22.2.pom
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-hdfs/2.6.5/hadoop-hdfs-2.6.5.pom
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-mapreduce-client-jobclient/2.6.5/hadoop-mapre

Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-auth/2.6.5/hadoop-auth-2.6.5.pom
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/hk2-utils/2.4.0-b34/hk2-utils-2.4.0-b34.pom
Downloaded https://repo1.maven.org/maven2/org/mortbay/jetty/jetty-util/6.1.26/jetty-util-6.1.26.pom
Downloaded https://repo1.maven.org/maven2/org/objenesis/objenesis/2.5.1/objenesis-2.5.1.pom
Downloaded https://repo1.maven.org/maven2/org/json4s/json4s-ast_2.12/3.5.3/json4s-ast_2.12-3.5.3.pom
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-mapreduce-client-shuffle/2.6.5/hadoop-mapreduce-client-shuffle-2.6.5.pom
Downloaded https://repo1.maven.org/maven2/com/google/code/gson/gson/2.2.4/gson-2.2.4.pom
Downloaded https://repo1.maven.org/maven2/org/eclipse/jetty/jetty-util/9.4.20.v20190813/jetty-util-9.4.20.v20190813.pom
Downloaded https://repo1.maven.org/maven2/commons-httpclient/commons-httpclient/3.1/commons-httpclient-3.1.pom
Downloaded https://repo1.maven.org/maven2/o

Downloaded https://repo1.maven.org/maven2/org/apache/apache/3/apache-3.pom
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-yarn-server/2.6.5/hadoop-yarn-server-2.6.5.pom
Downloaded https://repo1.maven.org/maven2/org/apache/httpcomponents/httpcomponents-client/4.2.5/httpcomponents-client-4.2.5.pom
Downloaded https://repo1.maven.org/maven2/org/apache/directory/server/apacheds-parent/2.0.0-M15/apacheds-parent-2.0.0-M15.pom
Downloaded https://repo1.maven.org/maven2/org/apache/httpcomponents/project/6/project-6.pom
Downloaded https://repo1.maven.org/maven2/org/apache/directory/project/project/31/project-31.pom
Downloaded https://repo1.maven.org/maven2/org/apache/apache/11/apache-11.pom
Downloaded https://repo1.maven.org/maven2/org/apache/directory/api/api-util/1.0.0-M20/api-util-1.0.0-M20.pom
Downloaded https://repo1.maven.org/maven2/commons-beanutils/commons-beanutils/1.7.0/commons-beanutils-1.7.0.pom
Downloaded https://repo1.maven.org/maven2/jline/jline/0.9.94/jline-0.9

Downloaded https://repo1.maven.org/maven2/xmlenc/xmlenc/0.52/xmlenc-0.52.jar
Downloaded https://repo1.maven.org/maven2/com/twitter/chill_2.12/0.9.3/chill_2.12-0.9.3.jar
Downloaded https://repo1.maven.org/maven2/commons-net/commons-net/3.1/commons-net-3.1.jar
Downloaded https://repo1.maven.org/maven2/commons-cli/commons-cli/1.2/commons-cli-1.2.jar
Downloaded https://repo1.maven.org/maven2/org/apache/xbean/xbean-asm6-shaded/4.8/xbean-asm6-shaded-4.8.jar
Downloaded https://repo1.maven.org/maven2/org/apache/avro/avro/1.8.2/avro-1.8.2.jar
Downloaded https://repo1.maven.org/maven2/com/esotericsoftware/kryo-shaded/4.0.2/kryo-shaded-4.0.2.jar
Downloaded https://repo1.maven.org/maven2/commons-io/commons-io/2.4/commons-io-2.4.jar
Downloaded https://repo1.maven.org/maven2/commons-lang/commons-lang/2.6/commons-lang-2.6.jar
Downloaded https://repo1.maven.org/maven2/commons-httpclient/commons-httpclient/3.1/commons-httpclient-3.1.jar
Downloaded https://repo1.maven.org/maven2/commons-beanutils/common

Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-common/2.6.5/hadoop-common-2.6.5.jar
Downloaded https://repo1.maven.org/maven2/org/codehaus/jackson/jackson-xc/1.9.13/jackson-xc-1.9.13.jar
Downloaded https://repo1.maven.org/maven2/javax/xml/stream/stax-api/1.0-2/stax-api-1.0-2.jar
Downloaded https://repo1.maven.org/maven2/aopalliance/aopalliance/1.0/aopalliance-1.0.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-yarn-client/2.6.5/hadoop-yarn-client-2.6.5.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-mapreduce-client-core/2.6.5/hadoop-mapreduce-client-core-2.6.5.jar
Downloaded https://repo1.maven.org/maven2/com/google/inject/guice/3.0/guice-3.0.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-mapreduce-client-shuffle/2.6.5/hadoop-mapreduce-client-shuffle-2.6.5.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-yarn-server-common/2.6.5/hadoop-yarn-server-common-2.6.5.jar
Downloaded

Downloaded https://repo1.maven.org/maven2/org/roaringbitmap/RoaringBitmap/0.7.45/RoaringBitmap-0.7.45.jar
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/external/aopalliance-repackaged/2.4.0-b34/aopalliance-repackaged-2.4.0-b34.jar
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/hk2-locator/2.4.0-b34/hk2-locator-2.4.0-b34.jar
Downloaded https://repo1.maven.org/maven2/org/javassist/javassist/3.18.1-GA/javassist-3.18.1-GA.jar
Downloaded https://repo1.maven.org/maven2/org/glassfish/jersey/media/jersey-media-jaxb/2.22.2/jersey-media-jaxb-2.22.2.jar
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/hk2-api/2.4.0-b34/hk2-api-2.4.0-b34.jar
Downloaded https://repo1.maven.org/maven2/org/glassfish/jersey/core/jersey-client/2.22.2/jersey-client-2.22.2.jar
Downloaded https://repo1.maven.org/maven2/org/scala-lang/modules/scala-xml_2.12/1.0.6/scala-xml_2.12-1.0.6.jar
Downloaded https://repo1.maven.org/maven2/org/json4s/json4s-scalap_2.12/3.5.3/json4s-scalap_2.12-3.

Downloaded https://repo1.maven.org/maven2/org/apache/parquet/parquet-format/2.4.0/parquet-format-2.4.0.jar
Downloaded https://repo1.maven.org/maven2/org/apache/orc/orc-core/1.5.5/orc-core-1.5.5-nohive.jar
Downloaded https://repo1.maven.org/maven2/org/apache/parquet/parquet-encoding/1.10.1/parquet-encoding-1.10.1.jar
Downloaded https://repo1.maven.org/maven2/org/glassfish/jersey/core/jersey-server/2.22.2/jersey-server-2.22.2.jar
Downloaded https://repo1.maven.org/maven2/commons-configuration/commons-configuration/1.6/commons-configuration-1.6.jar
Downloaded https://repo1.maven.org/maven2/com/google/protobuf/protobuf-java/2.5.0/protobuf-java-2.5.0.jar
Downloaded https://repo1.maven.org/maven2/org/spark-project/spark/unused/1.0.0/unused-1.0.0.jar
Downloaded https://repo1.maven.org/maven2/io/airlift/aircompressor/0.10/aircompressor-0.10.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-core_2.12/2.4.3/spark-core_2.12-2.4.3.jar
Downloaded https://repo1.maven.org/maven2/or

Downloaded https://repo1.maven.org/maven2/org/apache/avro/avro/1.8.2/avro-1.8.2-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/xbean/xbean-asm6-shaded/4.8/xbean-asm6-shaded-4.8-sources.jar
Downloaded https://repo1.maven.org/maven2/org/codehaus/jackson/jackson-mapper-asl/1.9.13/jackson-mapper-asl-1.9.13-sources.jar
Downloaded https://repo1.maven.org/maven2/commons-cli/commons-cli/1.2/commons-cli-1.2-sources.jar
Downloaded https://repo1.maven.org/maven2/xmlenc/xmlenc/0.52/xmlenc-0.52-sources.jar
Downloaded https://repo1.maven.org/maven2/org/objenesis/objenesis/2.5.1/objenesis-2.5.1-sources.jar
Downloaded https://repo1.maven.org/maven2/commons-io/commons-io/2.4/commons-io-2.4-sources.jar
Downloaded https://repo1.maven.org/maven2/log4j/log4j/1.2.17/log4j-1.2.17-sources.jar
Downloaded https://repo1.maven.org/maven2/commons-codec/commons-codec/1.10/commons-codec-1.10-sources.jar
Downloaded https://repo1.maven.org/maven2/commons-digester/commons-digester/1.8/commons-digester

Downloaded https://repo1.maven.org/maven2/com/twitter/chill_2.12/0.9.3/chill_2.12-0.9.3-sources.jar
Downloaded https://repo1.maven.org/maven2/org/mortbay/jetty/jetty-util/6.1.26/jetty-util-6.1.26-sources.jar
Downloaded https://repo1.maven.org/maven2/xerces/xercesImpl/2.9.1/xercesImpl-2.9.1-sources.jar
Downloaded https://repo1.maven.org/maven2/commons-beanutils/commons-beanutils/1.7.0/commons-beanutils-1.7.0-sources.jar
Downloaded https://repo1.maven.org/maven2/org/xerial/snappy/snappy-java/1.1.7.3/snappy-java-1.1.7.3-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-common/2.6.5/hadoop-common-2.6.5-sources.jar
Downloaded https://repo1.maven.org/maven2/xml-apis/xml-apis/1.3.04/xml-apis-1.3.04-sources.jar
Downloaded https://repo1.maven.org/maven2/org/codehaus/jackson/jackson-jaxrs/1.9.13/jackson-jaxrs-1.9.13-sources.jar
Downloaded https://repo1.maven.org/maven2/com/univocity/univocity-parsers/2.7.3/univocity-parsers-2.7.3-sources.jar
Downloaded https://repo1.

Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-unsafe_2.12/2.4.3/spark-unsafe_2.12-2.4.3-sources.jar
Downloaded https://repo1.maven.org/maven2/javax/servlet/javax.servlet-api/3.1.0/javax.servlet-api-3.1.0-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-mapreduce-client-jobclient/2.6.5/hadoop-mapreduce-client-jobclient-2.6.5-sources.jar
Downloaded https://repo1.maven.org/maven2/org/lz4/lz4-java/1.4.0/lz4-java-1.4.0-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/httpcomponents/httpclient/4.2.5/httpclient-4.2.5-sources.jar
Downloaded https://repo1.maven.org/maven2/com/ning/compress-lzf/1.0.3/compress-lzf-1.0.3-sources.jar
Downloaded https://repo1.maven.org/maven2/com/github/luben/zstd-jni/1.3.2-2/zstd-jni-1.3.2-2-sources.jar
Downloaded https://repo1.maven.org/maven2/org/slf4j/jul-to-slf4j/1.7.16/jul-to-slf4j-1.7.16-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/commons/commons-lang3/3.5/commons-lang3-3.5-

Downloaded https://repo1.maven.org/maven2/javax/xml/stream/stax-api/1.0-2/stax-api-1.0-2-sources.jar
Downloaded https://repo1.maven.org/maven2/org/codehaus/janino/commons-compiler/3.0.9/commons-compiler-3.0.9-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/ivy/ivy/2.4.0/ivy-2.4.0-sources.jar
Downloaded https://repo1.maven.org/maven2/org/glassfish/hk2/hk2-utils/2.4.0-b34/hk2-utils-2.4.0-b34-sources.jar
Downloaded https://repo1.maven.org/maven2/org/scala-lang/modules/scala-parser-combinators_2.12/1.1.0/scala-parser-combinators_2.12-1.1.0-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/orc/orc-shims/1.5.5/orc-shims-1.5.5-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/orc/orc-core/1.5.5/orc-core-1.5.5-sources.jar
Downloaded https://repo1.maven.org/maven2/org/codehaus/janino/janino/3.0.9/janino-3.0.9-sources.jar
Downloaded https://repo1.maven.org/maven2/io/airlift/aircompressor/0.10/aircompressor-0.10-sources.jar
Downloaded https://repo1.m

Downloaded https://repo1.maven.org/maven2/org/apache/parquet/parquet-jackson/1.10.1/parquet-jackson-1.10.1-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/arrow/arrow-memory/0.10.0/arrow-memory-0.10.0-sources.jar
Downloaded https://repo1.maven.org/maven2/oro/oro/2.0.8/oro-2.0.8-sources.jar
Downloaded https://repo1.maven.org/maven2/com/fasterxml/jackson/module/jackson-module-paranamer/2.7.9/jackson-module-paranamer-2.7.9-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-sql_2.12/2.4.3/spark-sql_2.12-2.4.3-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/arrow/arrow-format/0.10.0/arrow-format-0.10.0-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-mapreduce-client-app/2.6.5/hadoop-mapreduce-client-app-2.6.5-sources.jar
Downloaded https://repo1.maven.org/maven2/xmlenc/xmlenc/0.52/xmlenc-0.52-sources.jar.sha1
Downloaded https://repo1.maven.org/maven2/xerces/xercesImpl/2.9.1/xercesImpl-2.9.1-sources.ja

import $ivy.$                                    

import com.databricks.spark.xml._

import $ivy.$                                  

import org.apache.spark.sql._

import org.apache.log4j._

In [2]:
Logger.getLogger("org").setLevel(Level.ERROR)

The path in the code below must point to the docker mounted volume not the host path.
The code below loads a local JAR file. This file contains XML processing code that was created for another chapter and is required later on. The key is the interp.load which in the Ammonite shell replaces the classPath.load method

In [3]:
val path=java.nio.file.FileSystems.getDefault.getPath("/home/jovyan/SparkDS/scalaspark/BookCode/aas/ch07-graph/target/ch07-graph-2.0.0-jar-with-dependencies.jar")
val x = ammonite.ops.Path(path)
interp.load.cp(x)

path: java.nio.file.Path = /home/jovyan/SparkDS/scalaspark/BookCode/aas/ch07-graph/target/ch07-graph-2.0.0-jar-with-dependencies.jar
x: os.Path = root/'home/'jovyan/'SparkDS/'scalaspark/'BookCode/'aas/"ch07-graph"/'target/"ch07-graph-2.0.0-jar-with-dependencies.jar"

In [4]:
import org.apache.hadoop.io.{Text, LongWritable}

import org.apache.hadoop.conf.Configuration

import org.apache.hadoop.io.{Text, LongWritable}


import org.apache.hadoop.conf.Configuration

In [5]:
def loadMedline(spark: SparkSession, path: String) = {
    import spark.implicits._  //Need to add in the external context to deal with conversion of Seq returned by the XML crawling library
    //@transient denotes that the instance shouldn't be serialized when the code is distributed
    @transient val conf = new Configuration()
    conf.set(com.databricks.spark.xml.XmlInputFormat.START_TAG_KEY,"<MedlineCitation ")
    conf.set(com.databricks.spark.xml.XmlInputFormat.END_TAG_KEY,"</MedlineCitation>")
    val sc = spark.sparkContext
    //Use the Hadoop file API. Recall that in all other examples we've read in the csv via sc.read.csv(). 
    val in = sc.newAPIHadoopFile(path, classOf[com.databricks.spark.xml.XmlInputFormat],
                                    classOf[org.apache.hadoop.io.LongWritable],classOf[Text],conf)
    in.map(line=>line._2.toString).toDS()
}

defined function loadMedline

In [6]:
val spark = new org.apache.spark.sql.almondinternals.NotebookSparkSessionBuilder().
            master("local[*]").config("spark.sql.shuffle.partitions",8).getOrCreate()
            

Loading spark-stubs


Downloaded https://repo1.maven.org/maven2/sh/almond/spark-stubs_24_2.12/0.7.2/spark-stubs_24_2.12-0.7.2.pom
Downloaded https://repo1.maven.org/maven2/sh/almond/spark-stubs_24_2.12/0.7.2/spark-stubs_24_2.12-0.7.2-sources.jar
Downloaded https://repo1.maven.org/maven2/sh/almond/spark-stubs_24_2.12/0.7.2/spark-stubs_24_2.12-0.7.2.jar


Getting spark JARs
Creating SparkSession


Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties


Spark UI

spark: SparkSession = org.apache.spark.sql.SparkSession@687464a

In [7]:
import spark.implicits._

import spark.implicits._

In [8]:
val medlineRaw = loadMedline(spark,"/home/jovyan/SparkDS/medline_data/")

medlineRaw: Dataset[String] = [value: string]

In [9]:
ops.pwd

res8: os.Path = root/'home/'jovyan/'SparkDS/'scalaspark

Scala's XML library needs to be explicitly imported.

In [10]:
import scala.xml._

import scala.xml._

In [11]:
val rawXML = medlineRaw.take(1)(0)
val elem = XML.loadString(rawXML)

take at cmd10.sc:1

1 / 1

rawXML: String = """<MedlineCitation Owner="NLM" Status="MEDLINE">
<PMID Version="1">10697468</PMID>
<DateCreated>
<Year>2000</Year>
<Month>03</Month>
<Day>06</Day>
</DateCreated>
<DateCompleted>
<Year>2000</Year>
<Month>03</Month>
<Day>06</Day>
</DateCompleted>
<DateRevised>
<Year>2013</Year>
<Month>11</Month>
<Day>21</Day>
</DateRevised>
<Article PubModel="Print">
<Journal>
<ISSN IssnType="Print">0099-2399</ISSN>
<JournalIssue CitedMedium="Print">
<Volume>1</Volume>
<Issue>6</Issue>
<PubDate>
<Year>1975</Year>
<Month>Jun</Month>
</PubDate>
</JournalIssue>
<Title>Journal of endodontics</Title>
<ISOAbbreviation>J Endod</ISOAbbreviation>
</Journal>
<ArticleTitle>Methodology and criteria in the evaluation of dental implants.</ArticleTitle>
<Pagination>
<MedlinePgn>193-202</MedlinePgn>
</Pagination>
<Abstract>
<AbstractText>This study was designed to develop an inexpensive, reproducible method for studying the reaction of the tissues of the oral cavity to the endosseo...
elem: Elem = <Med

In [12]:
elem.label

res11: String = "MedlineCitation"

In [13]:
elem \ "MeshHeadingList"

res12: NodeSeq = NodeSeq(
  <MeshHeadingList>
<MeshHeading>
<DescriptorName UI="D000818" MajorTopicYN="N">Animals</DescriptorName>
</MeshHeading>
<MeshHeading>
<DescriptorName UI="D003758" MajorTopicYN="N">Dental Implantation, Endosseous</DescriptorName>
<QualifierName UI="Q000009" MajorTopicYN="N">adverse effects</QualifierName>
<QualifierName UI="Q000379" MajorTopicYN="Y">methods</QualifierName>
</MeshHeading>
<MeshHeading>
<DescriptorName UI="D015921" MajorTopicYN="Y">Dental Implants</DescriptorName>
<QualifierName UI="Q000592" MajorTopicYN="N">standards</QualifierName>
</MeshHeading>
<MeshHeading>
<DescriptorName UI="D019232" MajorTopicYN="N">Dental Restoration Failure</DescriptorName>
</MeshHeading>
<MeshHeading>
<DescriptorName UI="D006168" MajorTopicYN="N">Guinea Pigs</DescriptorName>
</MeshHeading>
<MeshHeading>
<DescriptorName UI="D008297" MajorTopicYN="N">Male</DescriptorName>
</MeshHeading>
<MeshHeading>
<DescriptorName UI="D010043" MajorTopicYN="N">Outcome and Process Asses

Crawl the XML document and get get all those elements in the document that are listed as majortopics. Note the operator \\ which searches the Elem instance (XML document) for elements at any child level with the document. The use of the operator \ searches only the immediate children nodes. 

In [14]:
def majorTopics(record: String): Seq[String]={
    val elem = XML.loadString(record)
    val dn = elem \\ "DescriptorName"
    val mt = dn.filter(n=>(n \ "@MajorTopicYN").text=="Y")
    mt.map(n=>n.text)
}

defined function majorTopics

In [15]:
majorTopics(rawXML)

res14: Seq[String] = List("Dental Implants")

In [16]:
val medline = medlineRaw.map(majorTopics)
medline.cache()
medline.take(1)(0)

take at cmd15.sc:3

1 / 1

medline: Dataset[Seq[String]] = [value: array<string>]
res15_1: Dataset[Seq[String]] = [value: array<string>]
res15_2: Seq[String] = List("Dental Implants")

In [17]:
medline.first

first at cmd16.sc:1

1 / 1

res16: Seq[String] = List("Dental Implants")

In [18]:
medline.count()

count at cmd17.sc:1

32 / 32

count at cmd17.sc:1

1 / 1

res17: Long = 240000L

In [19]:
val topics=medline.flatMap(mesh=>mesh).toDF("topic")
topics.createOrReplaceTempView("topics")
val topicDist = spark.sql("""
select topic, count(*) as cnt
from topics
group by topic
order by cnt desc
""")
topicDist.show()

show at cmd18.sc:9

32 / 32

show at cmd18.sc:9

8 / 8

+--------------------+----+
|               topic| cnt|
+--------------------+----+
|            Research|1649|
|             Disease|1349|
|           Neoplasms|1123|
|        Tuberculosis|1066|
|       Public Policy| 816|
|       Jurisprudence| 796|
|          Demography| 763|
| Population Dynamics| 753|
|           Economics| 690|
|            Medicine| 682|
|Socioeconomic Fac...| 655|
|               Blood| 631|
|            Politics| 631|
|Emigration and Im...| 601|
|       Social Change| 577|
|          Physicians| 560|
|            Mutation| 542|
|   Abortion, Induced| 503|
|          Anesthesia| 483|
|       Public Health| 479|
+--------------------+----+
only showing top 20 rows



topics: DataFrame = [topic: string]
topicDist: DataFrame = [topic: string, cnt: bigint]

In [20]:
topicDist.count()

count at cmd19.sc:1

32 / 32

count at cmd19.sc:1

8 / 8

count at cmd19.sc:1

8 / 8

count at cmd19.sc:1

8 / 8

count at cmd19.sc:1

1 / 1

res19: Long = 14548L

In [21]:
val list = List(1,2,3)
val combs = list.reverse.combinations(2)
combs.foreach(println)

List(3, 2)
List(3, 1)
List(2, 1)


list: List[Int] = List(1, 2, 3)
combs: Iterator[List[Int]] = empty iterator

We are want to explore the co-occurences of major topics 

In [22]:
val topicPairs = medline.flatMap(t=> {
    t.sorted.combinations(2)
}).toDF("pairs")


topicPairs: DataFrame = [pairs: array<string>]

In [23]:
topicPairs.createOrReplaceTempView("topic_pairs")
val cooccurs = spark.sql("""
select pairs,count(*) as cnt
from topic_pairs
group by pairs
""")
cooccurs.cache()
cooccurs.count()

count at cmd22.sc:8

32 / 32

count at cmd22.sc:8

8 / 8

count at cmd22.sc:8

1 / 1

cooccurs: DataFrame = [pairs: array<string>, cnt: bigint]
res22_2: DataFrame = [pairs: array<string>, cnt: bigint]
res22_3: Long = 213745L

In [24]:
cooccurs.createOrReplaceTempView("cooccurs")
spark.sql("""
select pairs, cnt
from cooccurs
order by cnt desc
limit 10""").collect().foreach(println)

collect at cmd23.sc:2

8 / 8

[WrappedArray(Demography, Population Dynamics),288]
[WrappedArray(Government Regulation, Social Control, Formal),254]
[WrappedArray(Emigration and Immigration, Population Dynamics),230]
[WrappedArray(Acquired Immunodeficiency Syndrome, HIV Infections),220]
[WrappedArray(Antibiotics, Antitubercular, Dermatologic Agents),205]
[WrappedArray(Analgesia, Anesthesia),183]
[WrappedArray(Economics, Population Dynamics),181]
[WrappedArray(Analgesia, Anesthesia and Analgesia),179]
[WrappedArray(Anesthesia, Anesthesia and Analgesia),177]
[WrappedArray(Population Dynamics, Population Growth),174]


The above results show the occurence of pairs ( created in the spark-sql sentance).This is a very natural but brute force way to do it. It is however very limited in the analysis that can be done. We used basic frequency data which is effectively disregarding the relational data. Graph analysis can increase the level of insight one can have about the network of interconnections. 

## GraphX

Graph consists of VertexRDD[VD] which is syntactic sugar for RDD[(VertexID,VD)] and edge is and EdgeRDD[ED] (which Im guessing involves a source and destination vertex). VD and ED are any types you like. Vertex ID is a unique long that because fo the distributed nature is naturally done with a hashing algorithm. ENsuring that it is unique is a challenge. As the graph gets bigger the chance of a hash collision increases so better algorithm needed. 


In [25]:
import java.nio.charset.StandardCharsets
import java.security.MessageDigest

import java.nio.charset.StandardCharsets

import java.security.MessageDigest

In [26]:
def hashID(str: String): Long = {
    val bytes = java.security.MessageDigest.getInstance("MD5").
        digest(str.getBytes(java.nio.charset.StandardCharsets.UTF_8))
        (bytes(0) & 0xFFL) | 
        ((bytes(1) & 0xFFL) << 8) |
        ((bytes(2) & 0xFFL) << 16) |
        ((bytes(3) & 0xFFL) << 24) |
        ((bytes(4) & 0xFFL) << 32) |
        ((bytes(5) & 0xFFL) << 40) |
        ((bytes(6) & 0xFFL) << 48) |
        ((bytes(7) & 0xFFL) << 56) 
}

defined function hashID

This hash function will produce a unique ID. However, its not clear to me how that index can be located within the data partitions if we had to do some random access or join. 

In [27]:
import org.apache.spark.sql.Row
import org.apache.spark.sql.functions.countDistinct

import org.apache.spark.sql.Row

import org.apache.spark.sql.functions.countDistinct

In [28]:
topics.getClass()

res27: Class[T] = class org.apache.spark.sql.Dataset

In [29]:
val verticies = topics.map{ case Row(topic: String) =>  (hashID(topic), topic) }.toDF("hash","topic")

verticies: DataFrame = [hash: bigint, topic: string]

In [30]:
val uniqueHashes = verticies.agg(countDistinct("hash")).take(1)

take at cmd29.sc:1

32 / 32

take at cmd29.sc:1

8 / 8

take at cmd29.sc:1

1 / 1

uniqueHashes: Array[Row] = Array([14548])

The above counts show that there is no hash collisions.

In [31]:
topics.count()

count at cmd30.sc:1

32 / 32

count at cmd30.sc:1

1 / 1

res30: Long = 280464L

The double : causes the import function to look for library that has the current version of scala as the target. A single : in that place does consider this and just adds a library reference the exact version specified

In [32]:
import $ivy.`org.apache.spark::spark-graphx:2.4.4`

Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-graphx_2.12/2.4.4/spark-graphx_2.12-2.4.4.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-parent_2.12/2.4.4/spark-parent_2.12-2.4.4.pom
Downloaded https://repo1.maven.org/maven2/net/sourceforge/f2j/arpack_combined_all/0.1/arpack_combined_all-0.1.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-mllib-local_2.12/2.4.4/spark-mllib-local_2.12-2.4.4.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-tags_2.12/2.4.4/spark-tags_2.12-2.4.4.pom
Downloaded https://repo1.maven.org/maven2/com/github/fommil/netlib/core/1.1.2/core-1.1.2.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-core_2.12/2.4.4/spark-core_2.12-2.4.4.pom
Downloaded https://repo1.maven.org/maven2/com/github/fommil/netlib/parent/1.1/parent-1.1.pom
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-launcher_2.12/2.4.4/spark-launcher_2.12-2.4.4.pom
Downloaded https://repo1.mav

Downloaded https://repo1.maven.org/maven2/org/typelevel/machinist_2.12/0.6.1/machinist_2.12-0.6.1.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-network-common_2.12/2.4.4/spark-network-common_2.12-2.4.4.jar
Downloaded https://repo1.maven.org/maven2/commons-codec/commons-codec/1.9/commons-codec-1.9-sources.jar
Downloaded https://repo1.maven.org/maven2/org/spire-math/spire-macros_2.12/0.13.0/spire-macros_2.12-0.13.0.jar
Downloaded https://repo1.maven.org/maven2/org/slf4j/slf4j-api/1.7.16/slf4j-api-1.7.16-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-tags_2.12/2.4.4/spark-tags_2.12-2.4.4-sources.jar
Downloaded https://repo1.maven.org/maven2/net/sourceforge/f2j/arpack_combined_all/0.1/arpack_combined_all-0.1.jar
Downloaded https://repo1.maven.org/maven2/com/google/code/findbugs/jsr305/1.3.9/jsr305-1.3.9-sources.jar
Downloaded https://repo1.maven.org/maven2/org/apache/spark/spark-mllib-local_2.12/2.4.4/spark-mllib-local_2.12-2.4.4.jar
Dow

import $ivy.$                                     

In [33]:
import org.apache.spark.graphx._

import org.apache.spark.graphx._

In [34]:
//take in the pair of topics (Seq[_]) and hash each topics. Note would have to be modified if
//the graph was directed and we needed to ensure the direcionality.

val edges = cooccurs.map{ case Row(topics: Seq[_], cnt: Long) =>
            val ids = topics.map(_.toString).map(hashID).sorted
            Edge(ids(0),ids(1),cnt)
}

edges: Dataset[Edge[Long]] = [srcId: bigint, dstId: bigint ... 1 more field]

In [35]:
//Note that there is not deduplication here . The verticies contains the full list of 
//topics but the Graph creator method does the dedup. This is not the case with edges as this would preclude multi-graphs.

val vertexRDD = verticies.rdd.map{
            case Row(hash: Long, topic: String) => (hash, topic) 
}
//Instantiate the graph (which then gives access to the graph functionality)
val topicGraph = Graph(vertexRDD, edges.rdd)
topicGraph.cache()

vertexRDD: org.apache.spark.rdd.RDD[(Long, String)] = MapPartitionsRDD[112] at map at cmd34.sc:1
topicGraph: Graph[String, Long] = org.apache.spark.graphx.impl.GraphImpl@190a65e2
res34_2: Graph[String, Long] = org.apache.spark.graphx.impl.GraphImpl@190a65e2

In [36]:
verticies.count()

count at cmd35.sc:1

32 / 32

count at cmd35.sc:1

1 / 1

res35: Long = 280464L

In [37]:
topicGraph.vertices.count()

mapPartitions at VertexRDD.scala:356

8 / 8

map at cmd34.sc:1

32 / 32

fold at VertexRDDImpl.scala:90

32 / 32

res36: Long = 14548L

Notice the topicGraph vertices have been reduced down to the number of unique vertices. 

### Connected Components

connectedComponent uses an algorithm that seeks out the lowest valued index that a node is connected too. Note that this entails a lot of spark jobs and takes some time on this 15k node graph.

In [38]:
val connenctComponentGraph = topicGraph.connectedComponents()

mapPartitions at VertexRDDImpl.scala:247

32 / 32

mapPartitions at VertexRDDImpl.scala:247

32 / 32

mapPartitions at GraphImpl.scala:207

8 / 8

fold at VertexRDDImpl.scala:90

32 / 32

mapPartitions at VertexRDDImpl.scala:247

32 / 32

mapPartitions at VertexRDDImpl.scala:251

32 / 32

mapPartitions at GraphImpl.scala:207

8 / 8

fold at VertexRDDImpl.scala:90

32 / 32

mapPartitions at VertexRDDImpl.scala:251

32 / 32

mapPartitions at VertexRDDImpl.scala:247

32 / 32

mapPartitions at GraphImpl.scala:207

8 / 8

fold at VertexRDDImpl.scala:90

32 / 32

mapPartitions at VertexRDDImpl.scala:247

32 / 32

mapPartitions at VertexRDDImpl.scala:251

32 / 32

mapPartitions at GraphImpl.scala:207

8 / 8

fold at VertexRDDImpl.scala:90

32 / 32

mapPartitions at VertexRDDImpl.scala:251

32 / 32

mapPartitions at VertexRDDImpl.scala:247

32 / 32

mapPartitions at GraphImpl.scala:207

8 / 8

fold at VertexRDDImpl.scala:90

32 / 32

mapPartitions at VertexRDDImpl.scala:247

32 / 32

mapPartitions at VertexRDDImpl.scala:251

32 / 32

mapPartitions at GraphImpl.scala:207

8 / 8

fold at VertexRDDImpl.scala:90

32 / 32

mapPartitions at VertexRDDImpl.scala:251

32 / 32

mapPartitions at VertexRDDImpl.scala:247

32 / 32

mapPartitions at GraphImpl.scala:207

8 / 8

fold at VertexRDDImpl.scala:90

32 / 32

mapPartitions at VertexRDDImpl.scala:251

32 / 32

mapPartitions at VertexRDDImpl.scala:247

32 / 32

mapPartitions at GraphImpl.scala:207

8 / 8

fold at VertexRDDImpl.scala:90

32 / 32

connenctComponentGraph: Graph[VertexId, Long] = org.apache.spark.graphx.impl.GraphImpl@208346ed

In [39]:
connenctComponentGraph.getClass

res38: Class[T] = class org.apache.spark.graphx.impl.GraphImpl

In [40]:
topicGraph.getClass

res39: Class[T] = class org.apache.spark.graphx.impl.GraphImpl

In [41]:
val conDF = connenctComponentGraph.vertices.toDF("vid","cid")

conDF: DataFrame = [vid: bigint, cid: bigint]

In [42]:
conDF.take(10)

take at cmd41.sc:1

1 / 1

res41: Array[Row] = Array(
  [-8919150701840946063,-9218306090261648869],
  [-4044118614182780436,-9218306090261648869],
  [-6349302109162374113,-9218306090261648869],
  [781395735921866696,-9218306090261648869],
  [7036778678359965292,-9218306090261648869],
  [-6498082352254260633,-9218306090261648869],
  [954610265810015642,-9218306090261648869],
  [-5459672828122526211,-9218306090261648869],
  [-3298299725015287104,-9218306090261648869],
  [6454957809509266477,-9218306090261648869]
)

In [43]:
val conCounts = conDF.groupBy("cid").count()

conCounts: DataFrame = [cid: bigint, count: bigint]

In [44]:
conCounts.columns

res43: Array[String] = Array("cid", "count")

In [45]:
conCounts.orderBy($"count".desc).show()

show at cmd44.sc:1

32 / 32

show at cmd44.sc:1

8 / 8

+--------------------+-----+
|                 cid|count|
+--------------------+-----+
|-9218306090261648869|13610|
|-8193948242717911820|    5|
|-2062883918534425492|    4|
|-7685954109876710390|    3|
| 1765411469112156596|    3|
|-7016546051037489808|    3|
| 2742772755763603550|    3|
|-8679136035911620397|    3|
| -784187332742198415|    3|
|-1570577283199790912|    2|
|-3191983795676547449|    2|
| 1115577032605495238|    2|
|-2808543729408555812|    2|
| 8525283272801893592|    2|
|-1046815223728304871|    2|
|-2845580087620065567|    2|
| -895976636745178042|    2|
|-2565470797001804556|    2|
|-7738387475383617571|    2|
| 2615079146997623981|    2|
+--------------------+-----+
only showing top 20 rows



Do an inner join on the vertices of the topic graph and the connected graph to get the names of the topics in the various connected components. Seems there is an implicit join on the id (the hash). How sure how it knows that this is what the join should be done on. Perhaps the connectedCOmponentGraph derived from topic graphs uses pointers to the verticies in the original graph rather than deep copies of the topicGraph

In [46]:
val topicComponents = topicGraph.vertices.innerJoin(
    connenctComponentGraph.vertices) {
    (topic,name,componentId) => (name,componentId.toLong)
}.toDF("topic","cid")

topicComponents: DataFrame = [topic: bigint, cid: struct<_1: string, _2: bigint>]

In [47]:
topicComponents.take(10)

take at cmd46.sc:1

1 / 1

res46: Array[Row] = Array(
  [-8919150701840946063,[Carcinoma, Papillary,-9218306090261648869]],
  [-4044118614182780436,[Fluorides, Topical,-9218306090261648869]],
  [-6349302109162374113,[Social Adjustment,-9218306090261648869]],
  [781395735921866696,[Dietary Proteins,-9218306090261648869]],
  [7036778678359965292,[Cardiovascular Abnormalities,-9218306090261648869]],
  [-6498082352254260633,[Urobilinogen,-9218306090261648869]],
  [954610265810015642,[Haptophyta,-9218306090261648869]],
  [-5459672828122526211,[Image Enhancement,-9218306090261648869]],
  [-3298299725015287104,[Anal Sacs,-9218306090261648869]],
  [6454957809509266477,[Jaw Neoplasms,-9218306090261648869]]
)

In [48]:
topicComponents.where("cid._2==-2062883918534425492").show(truncate=false)

show at cmd47.sc:1

1 / 1

show at cmd47.sc:1

4 / 4

show at cmd47.sc:1

20 / 20

show at cmd47.sc:1

7 / 7

+--------------------+------------------------------------------------+
|topic               |cid                                             |
+--------------------+------------------------------------------------+
|-1870678893086276394|[Serotyping, -2062883918534425492]              |
|-1233269114313988317|[Campylobacter coli, -2062883918534425492]      |
|-2062883918534425492|[Campylobacter jejuni, -2062883918534425492]    |
|4763791955467795057 |[Campylobacter Infections, -2062883918534425492]|
+--------------------+------------------------------------------------+



### Degree Distribution 

In [49]:
val degrees: VertexRDD[Int] = topicGraph.degrees.cache()

degrees: VertexRDD[Int] = VertexRDDImpl[306] at RDD at VertexRDD.scala:57

In [50]:
degrees.map(_._2).stats()

mapPartitions at GraphImpl.scala:207

8 / 8

stats at cmd49.sc:1

32 / 32

res49: org.apache.spark.util.StatCounter = (count: 13721, mean: 31.155892, stdev: 65.497591, max: 2596.000000, min: 1.000000)

In [51]:
topicGraph.vertices.count()

fold at VertexRDDImpl.scala:90

32 / 32

res50: Long = 14548L

In [52]:
import org.apache.spark.sql.functions.countDistinct

import org.apache.spark.sql.functions.countDistinct

In [53]:
val singleTopics=medline.filter(x=>x.size==1).flatMap(x=>x).toDF("topic")

singleTopics: DataFrame = [topic: string]

In [54]:
singleTopics.createOrReplaceTempView("sing")
spark.sql("""
select distinct topic from sing
""").count()


count at cmd53.sc:2

32 / 32

count at cmd53.sc:2

8 / 8

count at cmd53.sc:2

1 / 1

res53_1: Long = 8243L

While these are the topics in the singletons it does not mean that these topics don't also occur in a topic pair. If they do occur in a topic pair then they aren't unconnected topics. We need to remove from the singleton list any topics to appear in a topic pair. Can use the DF except method here. 

In [55]:
val linkedSubset = topicPairs.flatMap(_.getAs[Seq[String]](0))
val singleTopics=medline.filter(x=>x.size==1).flatMap(x=>x)

linkedSubset: Dataset[String] = [value: string]
singleTopics: Dataset[String] = [value: string]

In [56]:
singleTopics.except(linkedSubset).count()

count at cmd55.sc:1

32 / 32

count at cmd55.sc:1

32 / 32

count at cmd55.sc:1

8 / 8

count at cmd55.sc:1

8 / 8

count at cmd55.sc:1

1 / 1

res55: Long = 827L

The 827 single topic documents that are unconnected accounts for the difference between thel medline vertex count and and the numver of vertices in the degree distribution graph. 

Lets look at the topics in the nodes that have the highest degree distribution. 

In [57]:
val topicDegrees = topicGraph.vertices.innerJoin(degrees){(topicID,name,degree)=>(name, degree.toInt)}.values.toDF("topic","degree")

topicDegrees: DataFrame = [topic: string, degree: int]

In [58]:
topicDegrees.orderBy($"degree".desc).show()

show at cmd57.sc:1

32 / 32

+-------------------+------+
|              topic|degree|
+-------------------+------+
|           Research|  2596|
|            Disease|  1746|
|          Neoplasms|  1202|
|              Blood|   914|
|       Pharmacology|   882|
|       Tuberculosis|   815|
|         Toxicology|   694|
|       Drug Therapy|   678|
|      Jurisprudence|   661|
|Biomedical Research|   633|
|         Physicians|   625|
|      Public Policy|   601|
|           Medicine|   590|
|         Metabolism|   578|
|Wounds and Injuries|   570|
|      Social Change|   570|
|              Brain|   569|
|          Hospitals|   557|
|              Urine|   551|
|          Economics|   548|
+-------------------+------+
only showing top 20 rows



### Chi-square test to detect statistically significant co-occurences of topics compared to a model where the topics are independently occuring the documents.

The previous analysis was done on the basis of frequency only and implicit in the interpretation of that data is that occurences are independant and their for frequency of occurence is meaningful in its own right. However, More sophosticated tests would use a null model and use that model to determine whether the frequency of occurence (or co-occurence) of topics was unexpectedly high relative to the null model (ie.e statistically significant). For categorical datat the natural choice is a chisquare tests which asserts that if the occurences are independant then in the limit the sampling distribution for the degrees of freedom we have will follow a chi square distribution. 

In [59]:
val T = medline.count() // This is the total number of co-occuring topics

count at cmd58.sc:1

32 / 32

count at cmd58.sc:1

1 / 1

T: Long = 240000L

The approach will be to create a new graph where a node is a topic and contains the number of occurences of that topic. Recalling that the edges of the topic graph counts the number of co-occurences of the topics in the same document. Collectivey this is all the data that is needed to calcuate the chi-squared statistic. 

In [60]:
import org.apache.spark.rdd.RDD

import org.apache.spark.rdd.RDD

In [61]:
val topicDistRdd = topicDist.map{
    case Row(topic: String, cnt: Long)=>(hashID(topic), cnt)
}.rdd

rdd at cmd60.sc:1

32 / 32

rdd at cmd60.sc:1

8 / 8

topicDistRdd: RDD[(Long, Long)] = MapPartitionsRDD[388] at rdd at cmd60.sc:1

In [62]:
val topicDistGraph = Graph(topicDistRdd, topicGraph.edges)

topicDistGraph: Graph[Long, Long] = org.apache.spark.graphx.impl.GraphImpl@1a990ff7

In [63]:
def chiSq(YY: Long, YB: Long, YA: Long, T: Long): Double = {
    val NB = T - YB
    val NA = T - YA
    val YN = YA - YY
    val NY = YB - YY
    val NN = T - NY - YN - YY
    val inner = math.abs(YY * NN - YN * NY) - T / 2.0
    T * math.pow(inner, 2) / (YA * NA * YB * NB)
}

defined function chiSq

EdgeTriplet[VD,ED] is a data structure that contains the attributes of the source and destination nodes and edge

In [64]:
val chiSquaredGraph = topicDistGraph.mapTriplets(triplet=>{
                            chiSq(triplet.attr,triplet.srcAttr,triplet.dstAttr, T)
                    })
chiSquaredGraph.edges.map(x=>x.attr).stats()

mapPartitions at VertexRDD.scala:356

8 / 8

rdd at cmd60.sc:1

8 / 8

rdd at cmd60.sc:1

8 / 8

mapPartitions at VertexRDDImpl.scala:247

8 / 8

stats at cmd63.sc:4

8 / 8

chiSquaredGraph: Graph[Long, Double] = org.apache.spark.graphx.impl.GraphImpl@11d8e4f7
res63_1: org.apache.spark.util.StatCounter = (count: 213745, mean: 877.956648, stdev: 5094.935171, max: 198668.408387, min: 0.000000)

The chisq calc calcualted the chisquared statistic for each pair of concepts. for a 2x2 contingency table the chisquared statistic for 99.999th percentile is 19.5 (ie. 0.001 % change that the null hyp of independance is false)

In [65]:
val interesting  = chiSquaredGraph.subgraph(
    triplet => triplet.attr>19.5)
interesting.edges.count()

fold at EdgeRDDImpl.scala:90

8 / 8

interesting: Graph[Long, Double] = org.apache.spark.graphx.impl.GraphImpl@f0918e6
res64_1: Long = 140575L

This shows that still there are 140575 edges that are significant. Not much of a reduction from the 240,000. 

### Triangle Count

GraphX has a triangle count algorithm built in. It returns a graph ( much like the when we used mapTriplets(ChiSqared()) to get the chisquared calculation

In [66]:
val triangleCount = interesting.triangleCount()

triangleCount: Graph[Int, Double] = org.apache.spark.graphx.impl.GraphImpl@334d8fdc

In [67]:
triangleCount.vertices.map(x=>x._2).stats()

map at GraphOps.scala:306

8 / 8

mapPartitions at GraphImpl.scala:207

8 / 8

mapPartitions at VertexRDD.scala:356

8 / 8

mapPartitions at VertexRDDImpl.scala:247

8 / 8

mapPartitions at GraphImpl.scala:207

8 / 8

stats at cmd66.sc:1

8 / 8

res66: org.apache.spark.util.StatCounter = (count: 14548, mean: 74.660159, stdev: 295.327094, max: 11023.000000, min: 0.000000)

In [68]:
val interestingDegrees = interesting.degrees.cache()

interestingDegrees: VertexRDD[Int] = VertexRDDImpl[468] at RDD at VertexRDD.scala:57

In [69]:
val maxTriangleGraph = interestingDegrees.mapValues(d=>d*(d-1)/2.0)

maxTriangleGraph: VertexRDD[Double] = VertexRDDImpl[470] at RDD at VertexRDD.scala:57

In [70]:
val clusterCoef = triangleCount.vertices.innerJoin(maxTriangleGraph) { (vertexId, triCount, maxTris) =>
                    {
                        if (maxTris==0.0) 0 else triCount/maxTris
                }
}

clusterCoef: VertexRDD[Double] = VertexRDDImpl[472] at RDD at VertexRDD.scala:57

In [78]:
clusterCoef.values.toDF("Coeff").orderBy($"Coeff".desc).describe().show()

describe at cmd77.sc:1

8 / 8

describe at cmd77.sc:1

8 / 8

describe at cmd77.sc:1

8 / 8

describe at cmd77.sc:1

1 / 1

+-------+-------------------+
|summary|              Coeff|
+-------+-------------------+
|  count|              13721|
|   mean|0.32470450645308857|
| stddev|0.30626890524041944|
|    min|                0.0|
|    max|                1.0|
+-------+-------------------+



In [79]:
clusterCoef.map(_._2).sum() / interesting.vertices.count()

sum at cmd78.sc:1

8 / 8

fold at VertexRDDImpl.scala:90

8 / 8

res78: Double = 0.3062462560518858

### Path Length

In this section we'll make use of the Pregel Operator included in GraphX. Pregel is based on a parallelisation framework that pre-dates mapreduce called BSP - Bulk Synchronous Parallel ( which Giraph uses)

In [83]:
// Takes two lookup tables that contain vertexId and distance and merges them. The merge operation is a union of vertices/distrances 
// with the shorted distance preserved 

def mergeMaps(m1: Map[VertexId, Int], m2: Map[VertexId, Int]): Map[VertexId, Int] = {
    def minThatExists(k:VertexId): Int = {
        math.min(m1.getOrElse(k, Int.MaxValue),
                m2.getOrElse(k, Int.MaxValue))
    }
    //the union of keys via the concatination of key sets
    (m1.keySet ++ m2.keySet).map {
        k=> (k, minThatExists(k))
    }.toMap
}
// Update merges the passed message into the vertex of the current node. This operation is exactly the same as the merge
//

def update(
            id: VertexId,
            state: Map[VertexId, Int],
            msg: Map[VertexId,Int]) = {
            mergeMaps(state,msg)
    }

defined function mergeMaps
defined function update

In [85]:
//This function will be used to iterate over the graph.It will take the message from one node increment the distances and pass it to 
//The neighbour b. It checks whether the updated data will cause the target vertice's lookup table to change and if so passes it to that neighbour.
//If not then it will pass an empty iterator indicating that no further updates will happen.
def checkIncrement(
    a: Map[VertexId,Int],
    b: Map[VertexId,Int],
    bid: VertexId)={
        
        val aplus = a.map {case (v,d) => v->(d+1) }
        if (b!=mergeMaps(aplus,b)){
            Iterator((bid,aplus))
        } else {
            Iterator.empty
        }
}

defined function checkIncrement

In [86]:
def iterate(e: EdgeTriplet[Map[VertexId,Int],_]) = {
    checkIncrement(e.srcAttr, e.dstAttr, e.dstId) ++
    checkIncrement(e.dstAttr, e.srcAttr, e.srcId)
}

defined function iterate